File: stock_search/evaluation/research.py
"""Run the research graph that produces evidence-backed analysis."""

from collections.abc import Iterable
from concurrent.futures import ThreadPoolExecutor
import os
import re
from typing import Any, Literal

from langchain.agents import create_agent
from langchain.agents.middleware import TodoListMiddleware
from langchain.tools import ToolRuntime, tool
from langchain_core.messages import HumanMessage
from langgraph.graph import END, START, StateGraph
from langgraph.store.memory import InMemoryStore
from llm_harness.clients import ChatOpenAI, WebLoaderAgent
from pydantic import BaseModel, Field

from ..utils import extract_domain
from .constants import ThresholdConfig

DEFAULT_ALLOWED_DOMAINS: tuple[str, ...] = (
    "example.com",
    "placeholder.org",
)

RESEARCH_AGENT_PROMPT = """Research with web search to find reliable, recent sources.
Focus on concrete facts, numbers, and named entities. No meta-language. Exclude ads/boilerplate.
Return concise bullet points with sources."""

WEBLOADER_AGENT_PROMPT = """Loads and summarizes pages from approved domains.
Extract only facts that help score moat, quality, and outlook. No meta-language.
Return short bullet points with sources if available."""

VALIDATOR_AGENT_PROMPT = """Validates claims for factual support.
Flag statements that lack evidence or conflict with sources. Return a short checklist of issues or "ok"."""

MAX_VALIDATION_RETRIES = 2
_STORE = InMemoryStore()


class ResearchGraphInput(BaseModel):
    """Represent the input payload for the research graph."""

    ticker: str


class ResearchGraphOutput(BaseModel):
    """Represent the output payload from the research graph."""

    structured_response: Any | None = None


class ValidationResult(BaseModel):
    """Represent the validator's decision about research completeness."""

    is_valid: bool = Field(description="True if the notes are grounded and factually supported.")
    reasons: list[str] = Field(default_factory=list, description="List of issues or reasons for invalidity.")


class ResearchGraphState(BaseModel):
    """Track research graph state."""

    ticker: str
    research_notes: str | None = None
    loaded_notes: str | None = None
    validation: ValidationResult | None = None
    attempts: int = 0
    plan: str | None = None
    structured_response: Any | None = None


def _get_allowed_domains() -> set[str]:
    """Get set of allowed domains from environment or defaults."""
    env = os.getenv("RESEARCH_ALLOWED_DOMAINS", "")
    if env:
        return {item.strip().lower() for item in env.split(",") if item.strip()}
    return set(ThresholdConfig.DEFAULT_ALLOWED_DOMAINS)


def _is_domain_allowed(domain: str, allowed: set[str]) -> bool:
    """Check if a domain or its parent is in the allowed list."""
    domain = domain.lower()
    return any(domain == entry or domain.endswith(f".{entry}") for entry in allowed)


def _filter_allowed_urls(urls: Iterable[str]) -> list[str]:
    """Filter URLs based on the allowed domains list."""
    allowed = _get_allowed_domains()
    return [url for url in urls if (d := extract_domain(url),) and _is_domain_allowed(d, allowed)]


def _extract_urls(text: str | None) -> list[str]:
    """Extract HTTP/HTTPS URLs from text."""
    if not text:
        return []
    return re.findall(r"https?://[^\s\]\)>,]+", text)


def _extract_todo_items(plan: str | None) -> list[str]:
    """Extract individual tasks from a markdown-style todo list."""
    if not plan:
        return []
    items = []
    for line in plan.splitlines():
        line = line.strip().lstrip("-").lstrip("*").strip()
        if not line:
            continue
        if line.lower().startswith(("todo:", "plan:")):
            line = line.split(":", 1)[-1].strip()
        items.append(line)
    return items


def _extract_ticker(text: str | None) -> str | None:
    """Extract a ticker symbol from text or a 'Ticker: SYMBOL' pattern."""
    if not text:
        return None
    candidate = text.strip()
    if re.fullmatch(r"[A-Za-z0-9\.-]{1,10}", candidate):
        return candidate.upper()
    match = re.search(r"Ticker:\s*([A-Za-z0-9\.-]+)", text)
    return match.group(1).upper() if match else None


def _chunk_list(values: list[str], size: int) -> list[list[str]]:
    """Split a list into chunks of a specified size."""
    if size <= 0:
        return [values]
    return [values[i : i + size] for i in range(0, len(values), size)]


@tool("load_research_memory", description="Load cached research notes for a ticker.")
def load_research_memory(
    ticker: str,
    runtime: ToolRuntime,
) -> str:
    """Load the persisted research-memory cache."""
    stored = runtime.store.get(
        ("research",),
        ticker.upper(),
    )
    return str(stored.value) if stored and stored.value else ""


@tool("save_research_memory", description="Save research notes for a ticker.")
def save_research_memory(
    ticker: str,
    notes: str,
    runtime: ToolRuntime,
) -> str:
    """Save research memory."""
    runtime.store.put(
        ("research",),
        ticker.upper(),
        notes,
    )
    return "saved"


class ResearchAgents:
    """Container for research-specific agents."""

    def __init__(self, system_prompt: str, response_format: type[BaseModel]):
        """Initialize the in-memory research writer state."""
        quality_model = os.getenv("QUALITY_LLM")
        fast_model = os.getenv("FAST_LLM")

        self.supervisor = create_agent(
            model=ChatOpenAI(
                model=quality_model,
                temperature=0,
                reasoning_effort="medium",
            ),
            tools=[
                load_research_memory,
                save_research_memory,
            ],
            system_prompt=system_prompt,
            response_format=response_format,
            middleware=[TodoListMiddleware()],
            store=_STORE,
        )

        self.searcher = create_agent(
            model=ChatOpenAI(
                model=quality_model,
                temperature=0,
                reasoning_effort="medium",
                web_search=True,
                web_search_max_results=ThresholdConfig.WEB_SEARCH_MAX_RESULTS,
            ),
            tools=[load_research_memory],
            system_prompt=RESEARCH_AGENT_PROMPT,
            store=_STORE,
        )

        self.loader = WebLoaderAgent(
            model=fast_model,
            temperature=0,
            reasoning_effort="low",
            system_prompt=WEBLOADER_AGENT_PROMPT,
        )

        self.validator = create_agent(
            model=ChatOpenAI(
                model=fast_model,
                temperature=0,
                reasoning_effort="medium",
            ),
            tools=[],
            system_prompt=VALIDATOR_AGENT_PROMPT,
            response_format=ValidationResult,
            store=_STORE,
        )


def _build_research_graph(system_prompt: str, response_format: type[BaseModel]) -> Any:
    """Build the LangGraph pipeline used for research reports."""
    agents = ResearchAgents(system_prompt, response_format)

    def planner_node(state: ResearchGraphState) -> dict[str, Any]:
        """Plan the next research tasks for the graph run."""
        prompt = f"Create a short todo list for researching this ticker. Do not answer the task.\nTicker: {state.ticker}"
        response = agents.supervisor.invoke(
            {
                "messages": [HumanMessage(content=prompt)],
            }
        )
        return {
            "plan": response["messages"][-1].content,
        }

    def fanout_node(state: ResearchGraphState) -> dict[str, Any]:
        """Expand the planner output into individual search tasks."""
        return {
            "ticker": state.ticker,
            "research_notes": None,
            "loaded_notes": None,
            "validation": None,
        }

    def websearch_node(state: ResearchGraphState) -> dict[str, Any]:
        """Run web searches for the pending research tasks."""
        ticker = state.ticker or _extract_ticker(state.ticker)
        queries = _extract_todo_items(state.plan) or [state.ticker]

        def _search_one(query: str) -> str:
            """Run one research search task against the web search backend."""
            prompt = f"{query}\nSummarize with concrete facts, numbers, and named entities. No meta-language.\nReturn concise bullet points. Include a 'Sources:' list with URLs you used."
            response = agents.searcher.invoke(
                {
                    "messages": [HumanMessage(content=prompt)],
                }
            )
            return response["messages"][-1].content

        with ThreadPoolExecutor(max_workers=min(len(queries), 5)) as executor:
            notes = "\n\n".join(executor.map(_search_one, queries))
        return {
            "ticker": ticker,
            "research_notes": notes,
        }

    def loader_node(state: ResearchGraphState) -> dict[str, Any]:
        """Load full article content for the selected sources."""
        urls = _filter_allowed_urls(_extract_urls(state.research_notes))
        if not urls:
            return {"loaded_notes": ""}

        url_batches = _chunk_list(urls, 5)

        def _load_batch(batch: list[str]) -> str:
            """Load one batch of selected research articles."""
            return agents.loader.invoke(f"Load and summarize these URLs:\n{batch}")

        with ThreadPoolExecutor(max_workers=min(len(url_batches), 5)) as executor:
            loaded = executor.map(_load_batch, url_batches)
            loaded = "\n\n".join(loaded)
        return {"loaded_notes": loaded}

    def validator_node(state: ResearchGraphState) -> dict[str, Any]:
        """Validate whether the gathered evidence is sufficient to write."""
        evidence = "\n\n".join(
            filter(
                None,
                [
                    state.research_notes,
                    state.loaded_notes,
                ],
            )
        )
        prompt = (
            f"Check the response for unsupported claims.\nIf everything is grounded, set is_valid=True. Otherwise set is_valid=False and list the reasons.\n\nNotes:\n{evidence}"
        )
        response = agents.validator.invoke({"messages": [HumanMessage(content=prompt)]})
        return {
            "validation": response.get("structured_response"),
            "attempts": state.attempts + 1,
        }

    def writer_node(state: ResearchGraphState) -> dict[str, Any]:
        """Draft the final research report from the validated evidence."""
        evidence = "\n\n".join(filter(None, [state.research_notes, state.loaded_notes]))
        prompt = f"{state.ticker}\n\nEvidence:\n{evidence}"
        response = agents.supervisor.invoke(
            {
                "messages": [HumanMessage(content=prompt)],
            }
        )
        result = response.get("structured_response")
        if state.ticker and result:
            _STORE.put(("research",), state.ticker, str(result))
        return {"structured_response": result}

    def route_from_validator(state: ResearchGraphState) -> Literal["writer", "end"]:
        """Choose the next graph edge after validation."""
        if state.validation and state.validation.is_valid:
            return "end"
        if state.attempts >= MAX_VALIDATION_RETRIES:
            return "end"
        return "writer"

    graph = StateGraph(
        ResearchGraphState,
        input_schema=ResearchGraphInput,
        output_schema=ResearchGraphOutput,
    )
    graph.add_node("plan", planner_node)
    graph.add_node("fanout", fanout_node)
    graph.add_node("websearch", websearch_node)
    graph.add_node("loader", loader_node)
    graph.add_node("validator", validator_node)
    graph.add_node("writer", writer_node)

    graph.add_edge(START, "plan")
    graph.add_edge("plan", "fanout")
    graph.add_edge("fanout", "websearch")
    graph.add_edge("fanout", "loader")
    graph.add_edge("websearch", "writer")
    graph.add_edge("loader", "writer")
    graph.add_edge("writer", "validator")
    graph.add_conditional_edges(
        "validator",
        route_from_validator,
        {"writer": "writer", "end": END},
    )

    return graph.compile()


def run_llm_evaluation(
    ticker: str,
    system_prompt: str,
    response_format: type[BaseModel],
) -> BaseModel:
    """Execute structured LLM search/analysis for a specific ticker."""
    graph = _build_research_graph(
        system_prompt,
        response_format,
    )
    response = graph.invoke(
        ResearchGraphInput(ticker=ticker),
    )
    return response.get("structured_response")


def save_research_graph_png(
    system_prompt: str,
    response_format: type[BaseModel],
    path: str,
) -> str | None:
    """Render the research graph to a PNG file."""
    graph = _build_research_graph(system_prompt, response_format)
    png_bytes = graph.get_graph().draw_mermaid_png()
    with open(path, "wb") as f:
        f.write(png_bytes)
    return path

File: src/stock-search/evaluation/research.ts
/** Run the research graph that produces evidence-backed analysis. */

import { writeFile } from "node:fs/promises";

import { HumanMessage } from "@langchain/core/messages";
import { type ToolRuntime, tool } from "@langchain/core/tools";
import {
	Annotation,
	END,
	InMemoryStore,
	START,
	StateGraph,
} from "@langchain/langgraph";
import { createAgent, todoListMiddleware } from "langchain";
import { WebLoaderAgent } from "llm-harness-js/agents";
import { ChatOpenAI } from "llm-harness-js/clients";
import { type ZodType, z } from "zod";

import { extractDomain } from "../utils.js";
import { ThresholdConfig } from "./constants.js";

const DEFAULT_ALLOWED_DOMAINS = ["example.com", "placeholder.org"] as const;
const RESEARCH_AGENT_PROMPT = `Research with web search to find reliable, recent sources.
Focus on concrete facts, numbers, and named entities. No meta-language. Exclude ads/boilerplate.
Return concise bullet points with sources.`;

const WEBLOADER_AGENT_PROMPT = `Loads and summarizes pages from approved domains.
Extract only facts that help score moat, quality, and outlook. No meta-language.
Return short bullet points with sources if available.`;

const VALIDATOR_AGENT_PROMPT = `Validates claims for factual support.
Flag statements that lack evidence or conflict with sources. Return a short checklist of issues or "ok".`;

const MAX_VALIDATION_RETRIES = 2;
const RESEARCH_STORE = new InMemoryStore();

const ValidationResultSchema = z.object({
	is_valid: z.boolean(),
	reasons: z.array(z.string()).default([]),
});

const researchGraphState = Annotation.Root({
	ticker: Annotation<string>,
	researchNotes: Annotation<string | null>,
	loadedNotes: Annotation<string | null>,
	validation: Annotation<z.infer<typeof ValidationResultSchema> | null>,
	attempts: Annotation<number>,
	plan: Annotation<string | null>,
	structuredResponse: Annotation<unknown | null>,
});

type ResearchGraphState = typeof researchGraphState.State;
type ResearchStore = {
	get(
		namespace: string[],
		key: string,
	): Promise<{ value?: { notes?: string } } | null>;
	put(
		namespace: string[],
		key: string,
		value: { notes: string },
	): Promise<void>;
};

function researchMemoryKey(ticker: string): string {
	return `research:${ticker.toUpperCase()}`;
}

function getAllowedDomains(): Set<string> {
	const env = process.env.RESEARCH_ALLOWED_DOMAINS ?? "";
	if (env.trim()) {
		return new Set(
			env
				.split(",")
				.map((item) => item.trim().toLowerCase())
				.filter(Boolean),
		);
	}
	return new Set(DEFAULT_ALLOWED_DOMAINS);
}

function isDomainAllowed(domain: string, allowed: Set<string>): boolean {
	const normalizedDomain = domain.toLowerCase();
	return [...allowed].some(
		(entry) =>
			normalizedDomain === entry || normalizedDomain.endsWith(`.${entry}`),
	);
}

function filterAllowedUrls(urls: Iterable<string>): string[] {
	const allowed = getAllowedDomains();
	return [...urls].filter((url) => {
		const domain = extractDomain(url);
		return domain ? isDomainAllowed(domain, allowed) : false;
	});
}

function extractUrls(text: string | null | undefined): string[] {
	if (!text) {
		return [];
	}
	return text.match(/https?:\/\/[^\s\])>,]+/g) ?? [];
}

function extractTodoItems(plan: string | null | undefined): string[] {
	if (!plan) {
		return [];
	}
	const items: string[] = [];
	for (const rawLine of plan.split("\n")) {
		let line = rawLine
			.trim()
			.replace(/^[-*]\s*/, "")
			.trim();
		if (!line) {
			continue;
		}
		if (
			line.toLowerCase().startsWith("todo:") ||
			line.toLowerCase().startsWith("plan:")
		) {
			line = line.split(":", 2).at(-1)?.trim() ?? "";
		}
		if (line) {
			items.push(line);
		}
	}
	return items;
}

function extractTicker(text: string | null | undefined): string | null {
	if (!text) {
		return null;
	}
	const candidate = text.trim();
	if (/^[A-Za-z0-9.-]{1,10}$/.test(candidate)) {
		return candidate.toUpperCase();
	}
	const match = text.match(/Ticker:\s*([A-Za-z0-9.-]+)/);
	return match?.[1]?.toUpperCase() ?? null;
}

function chunkList(values: string[], size: number): string[][] {
	if (size <= 0) {
		return [values];
	}
	const chunks: string[][] = [];
	for (let index = 0; index < values.length; index += size) {
		chunks.push(values.slice(index, index + size));
	}
	return chunks;
}

function stringifyMessageContent(content: unknown): string {
	if (typeof content === "string") {
		return content;
	}
	return JSON.stringify(content);
}

function lastMessageText(response: unknown): string {
	const messages =
		response &&
		typeof response === "object" &&
		Array.isArray((response as { messages?: unknown[] }).messages)
			? (response as { messages: unknown[] }).messages
			: [];
	const lastMessage = messages.at(-1) as { content?: unknown } | undefined;
	return stringifyMessageContent(lastMessage?.content ?? "");
}

function structuredResponseValue<T>(response: unknown): T | null {
	const structuredResponse =
		response && typeof response === "object"
			? (response as { structuredResponse?: T }).structuredResponse
			: undefined;
	return structuredResponse ?? null;
}

function createResearchAgents(systemPrompt: string, responseFormat: ZodType) {
	const qualityModel = process.env.QUALITY_LLM;
	const fastModel = process.env.FAST_LLM;
	if (!qualityModel || !fastModel) {
		throw new Error("No model configured. Set FAST_LLM and QUALITY_LLM.");
	}

	const loadResearchMemory = tool(
		async ({ ticker }: { ticker: string }, runtime: ToolRuntime) => {
			const key = researchMemoryKey(ticker);
			const store =
				(runtime.store as unknown as ResearchStore | null) ?? RESEARCH_STORE;
			const stored = await store.get(["research"], key);
			const value = stored?.value?.notes;
			return typeof value === "string" ? value : "";
		},
		{
			name: "load_research_memory",
			description: "Load cached research notes for a ticker.",
			schema: z.object({
				ticker: z.string(),
			}),
		},
	);

	const saveResearchMemory = tool(
		async (
			{ ticker, notes }: { ticker: string; notes: string },
			runtime: ToolRuntime,
		) => {
			const key = researchMemoryKey(ticker);
			const store =
				(runtime.store as unknown as ResearchStore | null) ?? RESEARCH_STORE;
			await store.put(["research"], key, { notes });
			return "saved";
		},
		{
			name: "save_research_memory",
			description: "Save research notes for a ticker.",
			schema: z.object({
				ticker: z.string(),
				notes: z.string(),
			}),
		},
	);

	const supervisor = createAgent({
		model: ChatOpenAI({
			model: qualityModel,
			temperature: 0,
			reasoningEffort: "medium",
		}),
		tools: [loadResearchMemory, saveResearchMemory],
		systemPrompt,
		responseFormat,
		middleware: [todoListMiddleware()],
		store: RESEARCH_STORE,
	});

	const searcher = createAgent({
		model: ChatOpenAI({
			model: qualityModel,
			temperature: 0,
			reasoningEffort: "medium",
			webSearch: true,
			webSearchMaxResults: ThresholdConfig.WEB_SEARCH_MAX_RESULTS,
		}),
		tools: [loadResearchMemory],
		systemPrompt: RESEARCH_AGENT_PROMPT,
		store: RESEARCH_STORE,
	});

	const loader = new WebLoaderAgent({
		model: fastModel,
		temperature: 0,
		reasoningEffort: "low",
		systemPrompt: WEBLOADER_AGENT_PROMPT,
	});

	const validator = createAgent({
		model: ChatOpenAI({
			model: fastModel,
			temperature: 0,
			reasoningEffort: "medium",
		}),
		tools: [],
		systemPrompt: VALIDATOR_AGENT_PROMPT,
		responseFormat: ValidationResultSchema,
		store: RESEARCH_STORE,
	});

	return {
		supervisor,
		searcher,
		loader,
		validator,
	};
}

function buildResearchGraph(systemPrompt: string, responseFormat: ZodType) {
	const agents = createResearchAgents(systemPrompt, responseFormat);

	async function plannerNode(state: ResearchGraphState) {
		const prompt = `Create a short todo list for researching this ticker. Do not answer the task.\nTicker: ${state.ticker}`;
		const response = await agents.supervisor.invoke({
			messages: [new HumanMessage(prompt)],
		});
		return {
			plan: lastMessageText(response),
		};
	}

	async function fanoutNode(state: ResearchGraphState) {
		return {
			ticker: state.ticker,
			researchNotes: null,
			loadedNotes: null,
			validation: null,
		};
	}

	async function websearchNode(state: ResearchGraphState) {
		const ticker = state.ticker || extractTicker(state.ticker) || state.ticker;
		const queries = extractTodoItems(state.plan) || [state.ticker];
		const notes = (
			await Promise.all(
				queries.map(async (query) => {
					const prompt = `${query}\nSummarize with concrete facts, numbers, and named entities. No meta-language.\nReturn concise bullet points. Include a 'Sources:' list with URLs you used.`;
					const response = await agents.searcher.invoke({
						messages: [new HumanMessage(prompt)],
					});
					return lastMessageText(response);
				}),
			)
		).join("\n\n");

		return {
			ticker,
			researchNotes: notes,
		};
	}

	async function loaderNode(state: ResearchGraphState) {
		const urls = filterAllowedUrls(extractUrls(state.researchNotes));
		if (urls.length === 0) {
			return { loadedNotes: "" };
		}

		const urlBatches = chunkList(urls, 5);
		const loadedNotes = (
			await Promise.all(
				urlBatches.map((batch) =>
					agents.loader.invoke(`Load and summarize these URLs:\n${batch}`),
				),
			)
		).join("\n\n");
		return { loadedNotes };
	}

	async function validatorNode(state: ResearchGraphState) {
		const evidence = [state.researchNotes, state.loadedNotes]
			.filter(Boolean)
			.join("\n\n");
		const prompt = `Check the response for unsupported claims.\nIf everything is grounded, set is_valid=True. Otherwise set is_valid=False and list the reasons.\n\nNotes:\n${evidence}`;
		const response = await agents.validator.invoke({
			messages: [new HumanMessage(prompt)],
		});
		return {
			validation:
				structuredResponseValue<z.infer<typeof ValidationResultSchema>>(
					response,
				),
			attempts: state.attempts + 1,
		};
	}

	async function writerNode(state: ResearchGraphState) {
		const evidence = [state.researchNotes, state.loadedNotes]
			.filter(Boolean)
			.join("\n\n");
		const prompt = `${state.ticker}\n\nEvidence:\n${evidence}`;
		const response = await agents.supervisor.invoke({
			messages: [new HumanMessage(prompt)],
		});
		const result = structuredResponseValue<unknown>(response);
		if (state.ticker && result != null) {
			await RESEARCH_STORE.put(["research"], researchMemoryKey(state.ticker), {
				notes: JSON.stringify(result),
			});
		}
		return {
			structuredResponse: result,
		};
	}

	function routeFromValidator(state: ResearchGraphState) {
		if (state.validation?.is_valid) {
			return END;
		}
		if (state.attempts >= MAX_VALIDATION_RETRIES) {
			return END;
		}
		return "writer";
	}

	return new StateGraph(researchGraphState)
		.addNode("plan", plannerNode)
		.addNode("fanout", fanoutNode)
		.addNode("websearch", websearchNode)
		.addNode("loader", loaderNode)
		.addNode("validator", validatorNode)
		.addNode("writer", writerNode)
		.addEdge(START, "plan")
		.addEdge("plan", "fanout")
		.addEdge("fanout", "websearch")
		.addEdge("fanout", "loader")
		.addEdge("websearch", "writer")
		.addEdge("loader", "writer")
		.addEdge("writer", "validator")
		.addConditionalEdges("validator", routeFromValidator, {
			writer: "writer",
			[END]: END,
		})
		.compile();
}

/** Execute structured LLM search/analysis for a specific ticker. */
export async function runLlmEvaluation<T extends ZodType>(
	ticker: string,
	systemPrompt: string,
	responseFormat: T,
): Promise<import("zod").infer<T>> {
	const graph = buildResearchGraph(systemPrompt, responseFormat);
	const response = await graph.invoke({
		ticker,
		researchNotes: null,
		loadedNotes: null,
		validation: null,
		attempts: 0,
		plan: null,
		structuredResponse: null,
	});
	return response.structuredResponse as import("zod").infer<T>;
}

/** Render the research graph to a PNG file. */
export async function saveResearchGraphPng<T extends ZodType>(
	systemPrompt: string,
	responseFormat: T,
	path: string,
): Promise<string | null> {
	const graph = buildResearchGraph(systemPrompt, responseFormat);
	const pngBytes = await graph.getGraph().drawMermaidPng();
	await writeFile(path, Buffer.from(await pngBytes.arrayBuffer()));
	return path;
}
